src/test/util/setup_common.cpp

```diff
@@ -120,18 +119,39 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, const std::vecto
         arguments = Cat(arguments, G_TEST_COMMAND_LINE_ARGUMENTS());
     }
     util::ThreadRename("test");
-    fs::create_directories(m_path_root);
-    m_args.ForceSetArg("-datadir", fs::PathToString(m_path_root));
-    gArgs.ForceSetArg("-datadir", fs::PathToString(m_path_root));
     gArgs.ClearPathCache();
     {
         SetupServerArgs(*m_node.args);
+        m_node.args->AddArg("-testdatadir", "Custom data directory (default: <random_string>)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
         std::string error;
         if (!m_node.args->ParseParameters(arguments.size(), arguments.data(), error)) {
             m_node.args->ClearArgs();
             throw std::runtime_error{error};
         }
     }
+
+    if (!m_node.args->IsArgSet("-testdatadir")) {
+        // By default, the data directory has a random name
+        const std::string testdatadir{g_insecure_rand_ctx_temp_path.rand256().ToString().c_str()};
+        m_path_root = fs::temp_directory_path() / "test_common_" PACKAGE_NAME / testdatadir;
+    } else {
+        // Custom data directory
+        const std::string testdatadir{m_node.args->GetArg("-testdatadir").value()};
+        m_path_root = fs::path(testdatadir.c_str()) / "test_temp";
```

Comment:
Same here; let's avoid using the string pointer `c_str()`, and instead use `fs::PathFromString(path)`

Reply:
I can't get this to work.

Reply:
```c++
fs::path dir = m_node.args->GetPathArg("-testdatadir");
// todo: check if path is valid
m_path_root = dir / "test_temp";
```